docs/how-to/hip_cpp_language_extensions.rst

```diff
@@ -0,0 +1,874 @@
+access it from the host, its address and size need to be queried using
+:cpp:func:`hipGetSymbolAddress` and :cpp:func:`hipGetSymbolSize` and copied with
+:cpp:func:`hipMemcpyToSymbol` or :cpp:func:`hipMemcpyFromSymbol`.
+
+__constant__
+--------------------------------------------------------------------------------
+
+Variables marked with ``__constant__`` reside in device memory. Variables in
+that address space are routed through the constant cache, but that address space
+has a limited logical size.
+This memory space is read-only from within kernels and can only be set by the
+host before kernel execution.
+
+To get the best performance benefit, these variables need a special access
+pattern to benefit from the constant cache - the access has to be uniform within
+a warp, otherwise the accesses are serialized.
+
+The constant cache reduces the pressure on the other caches and may enable
+higher throughput and lower latency accesses.
+
+To set the ``__constant__`` variables the host must copy the data to the device
+using :cpp:func:`hipMemcpyToSymbol`, for example:
+
+.. code-block:: cpp
+
+    __constant__ int const_array[8];
+
+    void set_constant_memory(){
+      int host_data[8] {1,2,3,4,5,6,7,8};
+
+      hipMemcpyToSymbol(const_array, host_data, sizeof(int) * 8);
+
+      // call kernel that accesses const_array
+    }
+
+__shared__
+--------------------------------------------------------------------------------
+
+Variables marked with ``__shared__`` are only accessible by threads within the
+same block and have the lifetime of that block. It is usually backed by on-chip
+shared memory, providing fast access to all threads within a block, which makes
+it perfectly suited for sharing variables.
+
+Shared memory can be allocated statically within the kernel, but the size
+of it has to be known at compile time.
+
+In order to dynamically allocate shared memory during runtime, but before the
+kernel is launched, the variable has to be declared  ``extern``, and the kernel
+launch has to specify the needed amount of ``extern`` shared memory in the launch
+configuration. The statically allocated shared memory is allocated without this
+parameter.
+
+.. code-block:: cpp
+
+  #include <hip/hip_runtime.h>
+
+  extern __shared__ int shared_array[];
+
+  __global__ void kernel(){
+    // initialize shared memory
+    shared_array[threadIdx.x] = threadIdx.x;
+    // use shared memory
+  }
+
+  int main(){
+    //shared memory in this case depends on the configurable block size
+    constexpr int blockSize = 256;
+    constexpr int sharedMemSize = blockSize * sizeof(int);
+    constexpr int gridSize = 2;
+
+    kernel<<<dim3(gridSize), dim3(blockSize), sharedMemSize, 0>>>();
+  }
+
+__managed__
+--------------------------------------------------------------------------------
+
+Managed memory is a special qualifier, that makes the marked memory available on
+the device and on the host. For more details see :ref:`unified_memory`.
+
+__restrict__
```

Comment:
It is somewhat of a language extension, but also compiler specific/a hint to the compiler. However since CUDA requires this, I'm quite sure HIP also has to support it, so I'd say it has its place in this documentation